src/include/duckdb/common/types/string_type.hpp

```diff
@@ -33,10 +33,15 @@ struct string_t {
 			// zero initialize the prefix first
 			// this makes sure that strings with length smaller than 4 still have an equal prefix
 			memset(prefix, 0, PREFIX_LENGTH);
+			value_.data = nullptr;
 			if (length == 0) {
 				return;
 			}
 			// small string: inlined
+			/* Note: this appears to write out-of bounds on `prefix` if `length` > `PREFIX_LENGTH`
+			 but this is not the case because the `value_` union `inlined` char array directly
+			 follows it with 8 more chars to use for the string value.
+			 */
 			memcpy(prefix, data, length);
 			prefix[length] = '\0';
```

Comment:
This generates an error:

```
>>               stderr:
>>               f4d/external/duckdb/duckdb.hpp:1278:4: runtime error: index 7 out of bounds for type 'char [4]'
>>                   #0 0x7f7c5124edaf in duckdb::string_t::string_t(char const*, unsigned int) f4d/external/duckdb/duckdb.hpp:1278
>>                   #1 0x7f7c5124c36f in duckdb::string_t::string_t(char const*) f4d/external/duckdb/duckdb.hpp:1285
>>                   #2 0x7f7c51240d46 in __cxx_global_var_init.819 f4d/external/duckdb/duckdb.cpp:42285
>>                   #3 0x7f7c51241011 in _GLOBAL__sub_I_duckdb.cpp f4d/external/duckdb/duckdb.cpp
>>                   #4 0x7f7c55a56caa in call_init.part.0 /home/engshare/third-party2/glibc/2.26/src/glibc-2.26/elf/dl-init.c:72:3
>>                   #5 0x7f7c55a56dc1 in call_init /home/engshare/third-party2/glibc/2.26/src/glibc-2.26/elf/dl-init.c:30:6
>>                   #6 0x7f7c55a56dc1 in _dl_init /home/engshare/third-party2/glibc/2.26/src/glibc-2.26/elf/dl-init.c:120:5
>>                   #7 0x7f7c55a46d59  (/usr/local/fbcode/platform007/lib/ld.so+0xd59)
>>
>>               SUMMARY: UndefinedBehaviorSanitizer: out-of-bounds-index f4d/external/duckdb/duckdb.hpp:1278:4 in
```

Reply:
Would it be possible to somehow modify the code to avoid this failure or add a hint?

```
#if defined(__has_feature)
#if __has_feature(__address_sanitizer__)
        __attribute__((__no_sanitize__("out-of-bounds-index")))
#endif
#endif
```

Reply:
Similarly, AddOperator needed the following hint:

```
struct AddOperator {
        template <class TA, class TB, class TR> static inline TR Operation(TA left, TB right)
#if defined(__has_feature)
#if __has_feature(__address_sanitizer__)
        __attribute__((__no_sanitize__("signed-integer-overflow")))
#endif
#endif
  {
                return left + right;
        }
};
```

Reply:
> Would it be possible to somehow modify the code to avoid this failure or add a hint?
> 
> ```
> #if defined(__has_feature)
> #if __has_feature(__address_sanitizer__)
>         __attribute__((__no_sanitize__("out-of-bounds-index")))
> #endif
> #endif
> ```

That indeed seems like a false positive on the address sanitizer. I believe this should never generate undefined behavior/actual out of bounds accesses. The hint seems like a good idea to prevent it from triggering.

Reply:
Yeah will add the hints.